File: examples/tcp/tcp-bbr-example.cc
/*
 * Copyright (c) 2018-20 NITK Surathkal
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation;
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * Authors: Aarti Nandagiri <aarti.nandagiri@gmail.com>
 *          Vivek Jain <jain.vivek.anand@gmail.com>
 *          Mohit P. Tahiliani <tahiliani@nitk.edu.in>
 */

// This program simulates the following topology:
//
//           1000 Mbps           10Mbps          1000 Mbps
//  Sender -------------- R1 -------------- R2 -------------- Receiver
//              5ms               10ms               5ms
//
// The link between R1 and R2 is a bottleneck link with 10 Mbps. All other
// links are 1000 Mbps.
//
// This program runs by default for 100 seconds and creates a new directory
// called 'bbr-results' in the ns-3 root directory. The program creates one
// sub-directory called 'pcap' in 'bbr-results' directory (if pcap generation
// is enabled) and three .dat files.
//
// (1) 'pcap' sub-directory contains six PCAP files:
//     * bbr-0-0.pcap for the interface on Sender
//     * bbr-1-0.pcap for the interface on Receiver
//     * bbr-2-0.pcap for the first interface on R1
//     * bbr-2-1.pcap for the second interface on R1
//     * bbr-3-0.pcap for the first interface on R2
//     * bbr-3-1.pcap for the second interface on R2
// (2) cwnd.dat file contains congestion window trace for the sender node
// (3) throughput.dat file contains sender side throughput trace
// (4) queueSize.dat file contains queue length trace from the bottleneck link
//
// BBR algorithm enters PROBE_RTT phase in every 10 seconds. The congestion
// window is fixed to 4 segments in this phase with a goal to achieve a better
// estimate of minimum RTT (because queue at the bottleneck link tends to drain
// when the congestion window is reduced to 4 segments).
//
// The congestion window and queue occupancy traces output by this program show
// periodic drops every 10 seconds when BBR algorithm is in PROBE_RTT phase.

#include "ns3/applications-module.h"
#include "ns3/core-module.h"
#include "ns3/flow-monitor-module.h"
#include "ns3/internet-module.h"
#include "ns3/network-module.h"
#include "ns3/point-to-point-module.h"
#include "ns3/traffic-control-module.h"

using namespace ns3;

std::string dir;
uint32_t prev = 0;
Time prevTime = Seconds(0);

// Calculate throughput
static void
TraceThroughput(Ptr<FlowMonitor> monitor)
{
    FlowMonitor::FlowStatsContainer stats = monitor->GetFlowStats();
    auto itr = stats.begin();
    Time curTime = Now();
    std::ofstream thr(dir + "/throughput.dat", std::ios::out | std::ios::app);
    thr << curTime << " "
        << 8 * (itr->second.txBytes - prev) /
               (1000 * 1000 * (curTime.GetSeconds() - prevTime.GetSeconds()))
        << std::endl;
    prevTime = curTime;
    prev = itr->second.txBytes;
    Simulator::Schedule(Seconds(0.2), &TraceThroughput, monitor);
}

// Check the queue size
void
CheckQueueSize(Ptr<QueueDisc> qd)
{
    uint32_t qsize = qd->GetCurrentSize().GetValue();
    Simulator::Schedule(Seconds(0.2), &CheckQueueSize, qd);
    std::ofstream q(dir + "/queueSize.dat", std::ios::out | std::ios::app);
    q << Simulator::Now().GetSeconds() << " " << qsize << std::endl;
    q.close();
}

// Trace congestion window
static void
CwndTracer(Ptr<OutputStreamWrapper> stream, uint32_t oldval, uint32_t newval)
{
    *stream->GetStream() << Simulator::Now().GetSeconds() << " " << newval / 1448.0 << std::endl;
}

void
TraceCwnd(uint32_t nodeId, uint32_t socketId)
{
    AsciiTraceHelper ascii;
    Ptr<OutputStreamWrapper> stream = ascii.CreateFileStream(dir + "/cwnd.dat");
    Config::ConnectWithoutContext("/NodeList/" + std::to_string(nodeId) +
                                      "/$ns3::TcpL4Protocol/SocketList/" +
                                      std::to_string(socketId) + "/CongestionWindow",
                                  MakeBoundCallback(&CwndTracer, stream));
}

int
main(int argc, char* argv[])
{
    // Naming the output directory using local system time
    time_t rawtime;
    struct tm* timeinfo;
    char buffer[80];
    time(&rawtime);
    timeinfo = localtime(&rawtime);
    strftime(buffer, sizeof(buffer), "%d-%m-%Y-%I-%M-%S", timeinfo);
    std::string currentTime(buffer);

    std::string tcpTypeId = "TcpBbr";
    std::string queueDisc = "FifoQueueDisc";
    uint32_t delAckCount = 2;
    bool bql = true;
    bool enablePcap = false;
    Time stopTime = Seconds(100);

    CommandLine cmd(__FILE__);
    cmd.AddValue("tcpTypeId", "Transport protocol to use: TcpNewReno, TcpBbr", tcpTypeId);
    cmd.AddValue("delAckCount", "Delayed ACK count", delAckCount);
    cmd.AddValue("enablePcap", "Enable/Disable pcap file generation", enablePcap);
    cmd.AddValue("stopTime",
                 "Stop time for applications / simulation time will be stopTime + 1",
                 stopTime);
    cmd.Parse(argc, argv);

    queueDisc = std::string("ns3::") + queueDisc;

    Config::SetDefault("ns3::TcpL4Protocol::SocketType", StringValue("ns3::" + tcpTypeId));
    Config::SetDefault("ns3::TcpSocket::SndBufSize", UintegerValue(4194304));
    Config::SetDefault("ns3::TcpSocket::RcvBufSize", UintegerValue(6291456));
    Config::SetDefault("ns3::TcpSocket::InitialCwnd", UintegerValue(10));
    Config::SetDefault("ns3::TcpSocket::DelAckCount", UintegerValue(delAckCount));
    Config::SetDefault("ns3::TcpSocket::SegmentSize", UintegerValue(1448));
    Config::SetDefault("ns3::DropTailQueue<Packet>::MaxSize", QueueSizeValue(QueueSize("1p")));
    Config::SetDefault(queueDisc + "::MaxSize", QueueSizeValue(QueueSize("100p")));

    NodeContainer sender;
    NodeContainer receiver;
    NodeContainer routers;
    sender.Create(1);
    receiver.Create(1);
    routers.Create(2);

    // Create the point-to-point link helpers
    PointToPointHelper bottleneckLink;
    bottleneckLink.SetDeviceAttribute("DataRate", StringValue("10Mbps"));
    bottleneckLink.SetChannelAttribute("Delay", StringValue("10ms"));

    PointToPointHelper edgeLink;
    edgeLink.SetDeviceAttribute("DataRate", StringValue("1000Mbps"));
    edgeLink.SetChannelAttribute("Delay", StringValue("5ms"));

    // Create NetDevice containers
    NetDeviceContainer senderEdge = edgeLink.Install(sender.Get(0), routers.Get(0));
    NetDeviceContainer r1r2 = bottleneckLink.Install(routers.Get(0), routers.Get(1));
    NetDeviceContainer receiverEdge = edgeLink.Install(routers.Get(1), receiver.Get(0));

    // Install Stack
    InternetStackHelper internet;
    internet.Install(sender);
    internet.Install(receiver);
    internet.Install(routers);

    // Configure the root queue discipline
    TrafficControlHelper tch;
    tch.SetRootQueueDisc(queueDisc);

    if (bql)
    {
        tch.SetQueueLimits("ns3::DynamicQueueLimits", "HoldTime", StringValue("1000ms"));
    }

    tch.Install(senderEdge);
    tch.Install(receiverEdge);

    // Assign IP addresses
    Ipv4AddressHelper ipv4;
    ipv4.SetBase("10.0.0.0", "255.255.255.0");

    Ipv4InterfaceContainer i1i2 = ipv4.Assign(r1r2);

    ipv4.NewNetwork();
    Ipv4InterfaceContainer is1 = ipv4.Assign(senderEdge);

    ipv4.NewNetwork();
    Ipv4InterfaceContainer ir1 = ipv4.Assign(receiverEdge);

    // Populate routing tables
    Ipv4GlobalRoutingHelper::PopulateRoutingTables();

    // Select sender side port
    uint16_t port = 50001;

    // Install application on the sender
    BulkSendHelper source("ns3::TcpSocketFactory", InetSocketAddress(ir1.GetAddress(1), port));
    source.SetAttribute("MaxBytes", UintegerValue(0));
    ApplicationContainer sourceApps = source.Install(sender.Get(0));
    sourceApps.Start(Seconds(0.1));
    // Hook trace source after application starts
    Simulator::Schedule(Seconds(0.1) + MilliSeconds(1), &TraceCwnd, 0, 0);
    sourceApps.Stop(stopTime);

    // Install application on the receiver
    PacketSinkHelper sink("ns3::TcpSocketFactory", InetSocketAddress(Ipv4Address::GetAny(), port));
    ApplicationContainer sinkApps = sink.Install(receiver.Get(0));
    sinkApps.Start(Seconds(0.0));
    sinkApps.Stop(stopTime);

    // Create a new directory to store the output of the program
    dir = "bbr-results/" + currentTime + "/";
    std::string dirToSave = "mkdir -p " + dir;
    if (system(dirToSave.c_str()) == -1)
    {
        exit(1);
    }

    // The plotting scripts are provided in the following repository, if needed:
    // https://github.com/mohittahiliani/BBR-Validation/
    //
    // Download 'PlotScripts' directory (which is inside ns-3 scripts directory)
    // from the link given above and place it in the ns-3 root directory.
    // Uncomment the following three lines to generate plots for Congestion
    // Window, sender side throughput and queue occupancy on the bottleneck link.
    //
    // system (("cp -R PlotScripts/gnuplotScriptCwnd " + dir).c_str ());
    // system (("cp -R PlotScripts/gnuplotScriptThroughput " + dir).c_str ());
    // system (("cp -R PlotScripts/gnuplotScriptQueueSize " + dir).c_str ());

    // Trace the queue occupancy on the second interface of R1
    tch.Uninstall(routers.Get(0)->GetDevice(1));
    QueueDiscContainer qd;
    qd = tch.Install(routers.Get(0)->GetDevice(1));
    Simulator::ScheduleNow(&CheckQueueSize, qd.Get(0));

    // Generate PCAP traces if it is enabled
    if (enablePcap)
    {
        if (system((dirToSave + "/pcap/").c_str()) == -1)
        {
            exit(1);
        }
        bottleneckLink.EnablePcapAll(dir + "/pcap/bbr", true);
    }

    // Check for dropped packets using Flow Monitor
    FlowMonitorHelper flowmon;
    Ptr<FlowMonitor> monitor = flowmon.InstallAll();
    Simulator::Schedule(Seconds(0 + 0.000001), &TraceThroughput, monitor);

    Simulator::Stop(stopTime + TimeStep(1));
    Simulator::Run();
    Simulator::Destroy();

    return 0;
}

File: examples/tcp/tcp-bbr-example.py
# SPDX-License-Identifier: GPL-2.0-only
#
# Copyright (c) 2018-20 NITK Surathkal
# Copyright (c) 2025 CSPIT, CHARUSAT
#
# Authors:
#   Aarti Nandagiri <aarti.nandagiri@gmail.com>
#   Vivek Jain <jain.vivek.anand@gmail.com>
#   Mohit P. Tahiliani <tahiliani@nitk.edu.in>
#   Urval Kheni <kheniurval777@gmail.com>
#   Ritesh Patel <riteshpatel.ce@charusat.ac.in>

"""
Python port of tcp-bbr-example.cc

Simulates the following topology:

    Sender ---- R1 ---- R2 ---- Receiver
     1000Mbps    10Mbps   1000Mbps
       5ms        10ms       5ms

Outputs:
- cwnd.dat        (congestion window)
- throughput.dat  (throughput)
- queueSize.dat   (queue length)
"""

import argparse
import os
import time

try:
    from ns import ns
except ModuleNotFoundError:
    raise SystemExit(
        "Error: ns3 Python module not found. "
        "Bindings may not be enabled or PYTHONPATH is incorrect."
    )

# Global trace variables
prev_tx = 0
prev_time = None
throughput_file = None
queue_file = None
cwnd_file = None
outdir = ""


def trace_throughput_callback():
    """Throughput trace callback."""
    global prev_tx, prev_time, throughput_file

    stats = ns.cppyy.gbl.monitor_ptr.GetFlowStats()
    if stats and not stats.empty():
        flow_stats = stats.begin().__deref__().second
        cur_time = ns.Now()

        bits = 8 * (flow_stats.txBytes - prev_tx)
        us = (cur_time - prev_time).ToDouble(ns.Time.US)

        if us > 0:
            throughput_file.write(f"{cur_time.GetSeconds():.6g}s {bits / us:.6g} Mbps\n")
            throughput_file.flush()

        prev_time = cur_time
        prev_tx = flow_stats.txBytes


def check_queue_size_callback():
    """Queue length tracer."""
    global queue_file

    q = ns.cppyy.gbl.qdisc_ptr.GetCurrentSize().GetValue()
    queue_file.write(f"{ns.Simulator.Now().GetSeconds():.6g} {q}\n")
    queue_file.flush()


def trace_cwnd_callback(oldval, newval):
    """Congestion window tracer."""
    global cwnd_file

    cwnd_file.write(f"{ns.Simulator.Now().GetSeconds():.6g} {newval / 1448.0:.6g}\n")
    cwnd_file.flush()


# C++ helper functions for scheduling Python callbacks
ns.cppyy.cppdef(r"""
#include "ns3/simulator.h"
#include "ns3/flow-monitor-helper.h"
#include "ns3/queue-disc.h"
#include "ns3/node.h"
#include "ns3/config.h"

using namespace ns3;

Ptr<FlowMonitor> monitor_ptr;
Ptr<QueueDisc> qdisc_ptr;

// Helper to create event from Python callback without parameters
EventImpl* pythonMakeEvent(void (*f)())
{
    return MakeEvent(f);
}

// Helper to connect cwnd callback and create event
EventImpl* pythonConnectCwndTrace(void (*f)(uint32_t, uint32_t))
{
    Config::ConnectWithoutContext(
        "/NodeList/0/$ns3::TcpL4Protocol/SocketList/0/CongestionWindow",
        MakeCallback(f));
    return nullptr;
}
""")


def main():
    global throughput_file, queue_file, cwnd_file, outdir, prev_time
    global prev_tx

    parser = argparse.ArgumentParser(description="tcp-bbr-example (Python)")
    parser.add_argument(
        "--tcpTypeId",
        default="TcpBbr",
        help="Transport protocol: TcpNewReno, TcpBbr",
    )
    parser.add_argument("--delAckCount", type=int, default=2, help="Delayed ACK count")
    parser.add_argument("--enablePcap", action="store_true", help="Enable pcap traces")
    parser.add_argument("--stopTime", type=float, default=100.0, help="Simulation stop time")
    args = parser.parse_args()

    ts = time.strftime("%d-%m-%Y-%I-%M-%S")
    outdir = f"bbr-results/{ts}/"
    os.makedirs(outdir, exist_ok=True)

    tcpType = args.tcpTypeId
    queueDisc = "ns3::FifoQueueDisc"
    delAck = args.delAckCount
    enablePcap = args.enablePcap
    stopTime = ns.Seconds(args.stopTime)

    ns.Config.SetDefault(
        "ns3::TcpL4Protocol::SocketType",
        ns.StringValue("ns3::" + tcpType),
    )
    ns.Config.SetDefault("ns3::TcpSocket::SndBufSize", ns.UintegerValue(4194304))
    ns.Config.SetDefault("ns3::TcpSocket::RcvBufSize", ns.UintegerValue(6291456))
    ns.Config.SetDefault("ns3::TcpSocket::InitialCwnd", ns.UintegerValue(10))
    ns.Config.SetDefault("ns3::TcpSocket::DelAckCount", ns.UintegerValue(delAck))
    ns.Config.SetDefault("ns3::TcpSocket::SegmentSize", ns.UintegerValue(1448))
    ns.Config.SetDefault(
        "ns3::DropTailQueue<Packet>::MaxSize", ns.QueueSizeValue(ns.QueueSize("1p"))
    )
    ns.Config.SetDefault(
        queueDisc + "::MaxSize",
        ns.QueueSizeValue(ns.QueueSize("100p")),
    )

    sender = ns.NodeContainer()
    receiver = ns.NodeContainer()
    routers = ns.NodeContainer()
    sender.Create(1)
    receiver.Create(1)
    routers.Create(2)

    bottleneck = ns.PointToPointHelper()
    bottleneck.SetDeviceAttribute("DataRate", ns.StringValue("10Mbps"))
    bottleneck.SetChannelAttribute("Delay", ns.StringValue("10ms"))

    edge = ns.PointToPointHelper()
    edge.SetDeviceAttribute("DataRate", ns.StringValue("1000Mbps"))
    edge.SetChannelAttribute("Delay", ns.StringValue("5ms"))

    senderEdge = edge.Install(sender.Get(0), routers.Get(0))
    r1r2 = bottleneck.Install(routers.Get(0), routers.Get(1))
    receiverEdge = edge.Install(routers.Get(1), receiver.Get(0))

    inet = ns.InternetStackHelper()
    inet.Install(sender)
    inet.Install(receiver)
    inet.Install(routers)

    tch = ns.TrafficControlHelper()
    tch.SetRootQueueDisc(queueDisc)
    tch.SetQueueLimits("ns3::DynamicQueueLimits", "HoldTime", ns.StringValue("1000ms"))
    tch.Install(senderEdge)
    tch.Install(receiverEdge)

    ipv4 = ns.Ipv4AddressHelper()
    ipv4.SetBase(ns.Ipv4Address("10.0.0.0"), ns.Ipv4Mask("255.255.255.0"))

    _ = ipv4.Assign(r1r2)

    ipv4.NewNetwork()
    ipv4.Assign(senderEdge)

    ipv4.NewNetwork()
    ir1 = ipv4.Assign(receiverEdge)

    ns.Ipv4GlobalRoutingHelper.PopulateRoutingTables()

    port = 50001

    source = ns.BulkSendHelper(
        "ns3::TcpSocketFactory",
        ns.InetSocketAddress(ir1.GetAddress(1), port).ConvertTo(),
    )
    source.SetAttribute("MaxBytes", ns.UintegerValue(0))
    sourceApps = source.Install(sender.Get(0))
    sourceApps.Start(ns.Seconds(0.1))
    sourceApps.Stop(stopTime)

    sink = ns.PacketSinkHelper(
        "ns3::TcpSocketFactory",
        ns.InetSocketAddress(ns.Ipv4Address.GetAny(), port).ConvertTo(),
    )
    sinkApps = sink.Install(receiver.Get(0))
    sinkApps.Start(ns.Seconds(0))
    sinkApps.Stop(stopTime)

    if enablePcap:
        os.makedirs(os.path.join(outdir, "pcap"), exist_ok=True)

    throughput_file = open(os.path.join(outdir, "throughput.dat"), "w")
    queue_file = open(os.path.join(outdir, "queueSize.dat"), "w")
    cwnd_file = open(os.path.join(outdir, "cwnd.dat"), "w")

    if enablePcap:
        bottleneck.EnablePcapAll(os.path.join(outdir, "pcap", "bbr"), True)

    flowmon = ns.FlowMonitorHelper()
    monitor = flowmon.InstallAll()

    tch.Uninstall(routers.Get(0).GetDevice(1))
    qdc = tch.Install(routers.Get(0).GetDevice(1))

    prev_time = ns.Now()
    prev_tx = 0

    ns.cppyy.gbl.monitor_ptr = monitor
    ns.cppyy.gbl.qdisc_ptr = qdc.Get(0)

    # Define wrapper functions that reschedule themselves
    def throughput_tracer():
        """Periodic throughput tracer that reschedules itself."""
        trace_throughput_callback()
        event = ns.cppyy.gbl.pythonMakeEvent(throughput_tracer)
        ns.Simulator.Schedule(ns.Seconds(0.2), event)

    def queue_size_tracer():
        """Periodic queue size tracer that reschedules itself."""
        check_queue_size_callback()
        event = ns.cppyy.gbl.pythonMakeEvent(queue_size_tracer)
        ns.Simulator.Schedule(ns.Seconds(0.2), event)

    # Schedule initial events using pythonMakeEvent
    event = ns.cppyy.gbl.pythonMakeEvent(throughput_tracer)
    ns.Simulator.Schedule(ns.Seconds(0.000001), event)

    event = ns.cppyy.gbl.pythonMakeEvent(queue_size_tracer)
    ns.Simulator.ScheduleNow(event)

    # Connect cwnd trace after socket is created
    def connect_cwnd_trace():
        """Connect the cwnd trace callback."""
        ns.cppyy.gbl.pythonConnectCwndTrace(trace_cwnd_callback)

    event = ns.cppyy.gbl.pythonMakeEvent(connect_cwnd_trace)
    ns.Simulator.Schedule(ns.Seconds(0.1) + ns.MilliSeconds(1), event)

    ns.Simulator.Stop(stopTime + ns.TimeStep(1))
    ns.Simulator.Run()
    ns.Simulator.Destroy()

    throughput_file.close()
    queue_file.close()
    cwnd_file.close()


if __name__ == "__main__":
    main()
